includes/processors/class.llms.processor.course.data.php

```diff
@@ -43,107 +54,114 @@ class LLMS_Processor_Course_Data extends LLMS_Abstract_Processor {
 
 	/**
 	 * Maximum number of students allowed in a course
-	 * when enrollment is higher than this number
-	 * throttling the calculations will be delayed
+	 *
+	 * When enrollment is higher than this number
+	 * throttling the calculations will be delayed.
 	 *
 	 * @var int
 	 */
 	private $throttle_max_students;
 
 	/**
-	 * When a calculation is throttled based on the number of students
-	 * this will determine the frequency with which the query can run
-	 * Should be time in seconds
-	 * default is HOUR_IN_SECONDS * 4
+	 * Frequency of calculation process when the process is throttled.
 	 *
 	 * @var int
 	 */
 	private $throttle_frequency;
 
-	/**
-	 * Called when queue is emptied and process is complete
-	 *
-	 * @return   void
-	 * @since    3.15.0
-	 * @version  3.15.0
-	 */
-	protected function complete() {
-
-		parent::complete();
-		$this->set_data( 'last_run', time() );
-
-	}
-
 	/**
 	 * Action triggered to queue queries needed to make the calculation
 	 *
-	 * @param    int $course_id  WP Post ID of the course
-	 * @return   void
-	 * @since    3.15.0
-	 * @version  3.15.0
+	 * @since 3.15.0
+	 * @since [version] Add throttling by course in progress and adjust last_run calculation to be specific to the course.
+	 *               Improve performance of the student query by removing unneeded sort columns.
+	 *
+	 * @param int $course_id WP Post ID of the course.
+	 * @return void
 	 */
 	public function dispatch_calc( $course_id ) {
 
-		$this->log( sprintf( 'course data calculation dispatched for course %d', $course_id ) );
-
-		// cancel process in case it's currently running
-		$this->cancel_process();
-
-		$args = array(
-			'post_id'  => $course_id,
-			'statuses' => array( 'enrolled' ),
-			'page'     => 1,
-			'per_page' => 100,
+		$this->log( sprintf( 'Course data calculation dispatched for course %d.', $course_id ) );
+
+		/**
+		 * Filter the query arguments used when calculating course data
+		 *
+		 * @since [version]
+		 *
+		 * @param array                      $args      Query arguments passed to LLMS_Student_Query.
+		 * @param LLMS_Processor_Course_Data $processor Instance of the data processor class.
+		 */
+		$args = apply_filters(
+			'llms_data_processor_course_data_student_query_args',
+			array(
+				'post_id'  => $course_id,
+				'statuses' => array( 'enrolled' ),
+				'page'     => 1,
+				'per_page' => 100,
+				'sort'     => array(
+					'id' => 'ASC',
+				),
+			),
+			$this
 		);
 
-		// get total number of pages
+		// Get total number of pages.
 		$query = new LLMS_Student_Query( $args );
 
-		// only queue if we have students in the course
+		// Only queue if we have students in the course.
 		if ( $query->found_results ) {
 
-			// throttle dispatch?
-			if ( $this->maybe_throttle( $query->found_results ) ) {
+			// Throttle dispatch based on number of students or because the course is already processing.
+			if ( $this->maybe_throttle( $query->found_results, $course_id ) ) {
```

Comment:
If the course is already processing we actually can throttle _before_ any query.
I think is reasonable considering that if the action that triggers the `dispatch_calc` has been fired, and the course is already processing then there are enrolled students... and even if they're not it's something that it'll be addressed later. Meantime we've avoided an expensive query. What do you think?
If you agree with this, I would make a check on if `is_already_processing_course( $course_id )` before performing the query. Pseudo-code:
```php
$throttle = $this->is_already_processing_course( $course_id );
if ( ! $throttle ) {
	$query    = new LLMS_Student_Query( $args );
	$throttle = $this->maybe_throttle( $query->found_results, $course_id, false );
}
if ( $throttle ) { 
	do_throttle();
	return;
}
// no throttling.
if ( $query->found_results ) {
	// Add each page to the queue.
	// Save queue and dispatch the process.
}
```
The `false` as third parameter for `maybe_throttle()` method would be to avoid checking on if the course is already processing twice:
```php
	protected function maybe_throttle( $num_students, $course_id, $check_already_processing_course = true ) {

		$throttled = false;

		if ( $check_already_processing_course && $this->is_already_processing_course( $course_id ) ) {

			$throttled = true;

		} 
```

well this is the _idea_. What do you think?

Reply:
I like this idea, makes sense. Especially more so when checking the meta data to determine if the course's processing is locked. Much better.

I wrote the update a bit differently than your pseudo code here but I think it achieves the same end. Thoughts?